lib/evmone/analysis.cpp

```diff
@@ -71,6 +71,35 @@ code_analysis analyze(evmc_revision rev, const uint8_t* code, size_t code_size)
 
         block.gas_cost += opcode_info.gas_cost;
 
+        // Potentially optimize away heavy operations whose result is immediately discarded by POP.
+        // Only operations that take 1 element from the stack and put 1 element back are considered.
+        // This change speeds up the execution of certain Shanghai attack blocks
+        // that called EXTCODESIZE and immediately discarded it by POP.
+        if (code_pos != code_end && *code_pos == OP_POP && opcode_info.stack_req == 1 &&
+            opcode_info.stack_change == 0)
+        {
+            bool pure_and_const_cost = false;
+            switch (opcode)
+            {
+            case OP_BALANCE:
+            case OP_EXTCODESIZE:
+            case OP_EXTCODEHASH:
+            case OP_BLOCKHASH:
+            case OP_SLOAD:
+                pure_and_const_cost = true;
+            }
+
+            if (pure_and_const_cost)
+            {
+                // Warning: we can only replace discarded instruction with a no-op
+                // if it's pure (has no side effects) and its cost is constant.
+                // For example, an operation that can add an account to the touch set (EIP-161)
+                // cannot be considered pure and optimized away.
+                analysis.instrs.emplace_back(noop);
```

Comment:
Have you tried no pushing any instruction here?

Reply:
@chfast You're right, it's not necessary to push any instruction. I've updated the PR.